per-bucket random replica order #2957

yaojingguo opened the pull request:
<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3041)

<!-- Reviewable:end -->

yaojingguo committed:
per-bucket random replica order #2957

tamird commented:
@yaojingguo you don't need to close the PR and open a new one; if you update your branch, your PR will be updated.

yaojingguo commented:
@tamird, got it, sorry for the trouble.

yaojingguo commented:
There is a race failure TestReplicateRogueRemovedNode in CircleCI. But I have successfully run `./build/circle-local.sh` on my laptop. How can I proceed from here?

tamird commented:
I rebuilt it for you, green now.

lizhongz commented:
Hi @yaojingguo , I have also worked on this issue. Your implementation is good except two small pieces.
* The maxIndex parameter of randPerm method is not necessary.
* The rand.Perm method is *pseudo-random*, which means the results will be the same every time you call it with the same argument. As a result, you should set a seed before calling it.

yaojingguo committed:
remove maxIndex #2957

yaojingguo commented:
@lizhongz 

> The maxIndex parameter of randPerm method is not necessary.

I have originally thought that `len`'s cost is O(n). After I read [Go Slices: usage and internals](https://blog.golang.org/go-slices-usage-and-internals), I know that `len`'s cost is O(1).  So I have removed `maxIndex` according to your suggestion.

> The rand.Perm method is pseudo-random, which means the results will be the same every time you call it with the same argument. As a result, you should set a seed before calling it.

`pseudo-random` means that the sequence is generated for the same seed.  It does not mean that `the results will be the same every time you call it with the same argument`. The default rand source has been initialized with `rand.Seed(randutil.NewPseudoSeed())` in ` main.go`.  So I think that we don't need to do it again here.

lizhongz commented:
@yaojingguo Good to know that the seed has been set in main.go.

tschottdorf commented:
Could you write some tests? For testing `randPerm` it will be better if you pass the random number generator in as a parameter (and not use `rand.*` directly).

tschottdorf commented on kv/replica_slice.go:
```diff
@@ -103,7 +105,8 @@ func (rs replicaSlice) SortByCommonAttributePrefix(attrs []string) int {
 	}
 	topIndex := len(rs) - 1
-	for bucket := 0; bucket < len(attrs); bucket++ {
+	bucket := 0
```
could you back out this part of the change (doesn't seem like an improvement) or explain why you prefer it this way?
Also, remove the TODO on line 102.

yaojingguo committed:
Changes for randPerm

1. Remove TODO
2. Back out the change for bucket variable
3. Add test for randPerm

yaojingguo commented:
@tschottdorf, I have created a new commit according to your suggestions.

tamird commented on kv/replica_slice_test.go:
```diff
@@ -110,3 +117,25 @@ func TestReplicaSetMoveToFront(t *testing.T) {
+func TestReplicaSetRandPerm(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	rs := createReplicaSlice()
+	r := rand.New(rand.NewSource(randutil.NewPseudoSeed()))
```
don't you want a known seed for this test? `r := rand.New(rand.NewSource(0))`?

tschottdorf commented on kv/replica_slice_test.go:
```diff
@@ -110,3 +117,25 @@ func TestReplicaSetMoveToFront(t *testing.T) {
+	rs.randPerm(0, 2, r.Perm)
+	stores := getStores(rs)[3:5]
+	exp = exp[3:5]
+	if !reflect.DeepEqual(stores, exp) {
```
as @tamird suggested, get use a known seed and verify that you actually get rearranged replica sets (in particular guarding against one-offs: get an example in which the first and last item of the shuffled part of the slice are not the same).

yaojingguo committed:
Changes for part shuffle

1. Add tests to verify the ordering of the shuffled part
2. Use a proven algorithm for random permutation

yaojingguo commented:
@tamird and @tschottdorf, a new commit according to your suggestions. Now the slice shuffle allocates a new slice because it make the verification of the shuffled slice ordering easy. If allocating a new slice is disallowed, I can do the shuffle in place.

tamird commented:
There is no need to allocate a new slice in `randPerm`. If you need a copy, make the copy in your test.

yaojingguo committed:
in-place shuffle

yaojingguo commented:
@tamird, a new commit which does  `in-place shuffle` and avoids the allocation of a new slice.

tschottdorf commented on kv/replica_slice_test.go:
```diff
@@ -110,3 +116,32 @@ func TestReplicaSetMoveToFront(t *testing.T) {
+	rs := createReplicaSlice()
+	r := rand.New(rand.NewSource(0))
+	rs.randPerm(0, 4, r.Intn)
+	exp := []roachpb.StoreID{1, 2, 3, 4, 5}
```
can you add a test case in which `5` moves?

tschottdorf commented:
LGTM once the test case is added. The test failure likely requires a rebase.

yaojingguo committed:
simplify randPerm and add more tests

1. Move topIndex outside of randPerm
2. Remove lengh check from randPerm
3. Add verifyRandPermOrdering function
4. Add more tests

yaojingguo commented:
@tschottdorf , a new commit after:
 
1. A rebase is performed.
2. Add more tests as you have suggested and make some other minor changes. 

For details, check the commit message and the code.

tschottdorf commented:
Thanks @yaojingguo! Please squash everything (one easy way to do that is `git fetch origin && git reset --soft origin/master && git commit`) and then we can merge it.
